planetiler-core/src/main/java/com/onthegomap/planetiler/util/LoopLineMerger.java

```diff
@@ -0,0 +1,339 @@
+    factory = new GeometryFactory(precisionModel);
+    return this;
+  }
+
+  public LoopLineMerger setMinLength(double minLength) {
+    this.minLength = minLength;
+    return this;
+  }
+
+  public LoopLineMerger setLoopMinLength(double loopMinLength) {
+    this.loopMinLength = loopMinLength;
+    return this;
+  }
+
+  public void add(Geometry geometry) {
+    geometry.apply((GeometryComponentFilter) component -> {
+      if (component instanceof LineString lineString) {
+        input.add(lineString);
+      }
+    });
+  }
+
+  private void merge() {
+    for (var node : output) {
+      if (node.getEdges().size() == 2) {
+        Edge a = node.getEdges().getFirst();
+        Edge b = node.getEdges().get(1);
+        node.getEdges().clear();
+        List<Coordinate> coordinates = new ArrayList<>();
+        coordinates.addAll(a.coordinates.reversed());
+        coordinates.addAll(b.coordinates.subList(1, b.coordinates.size()));
+        Edge c = new Edge(a.to, b.to, coordinates, a.length + b.length);
+        a.to.removeEdge(a.reversed);
+        b.to.removeEdge(b.reversed);
+        a.to.addEdge(c);
+        b.to.addEdge(c.reversed);
+      }
+    }
+  }
+
+  private void removeLoops() {
+    for (var node : output) {
+      for (var edge : List.copyOf(node.getEdges())) {
+        var allPaths = findAllPaths(edge.from, edge.to, loopMinLength);
+        if (allPaths.size() > 1) {
+          for (var path : allPaths.subList(1, allPaths.size())) {
+            if (path.size() > 0) {
+              path.get(0).remove();
+            }
+          }
+        }
+      }
+    }
+  }
+
+  List<List<Edge>> findAllPaths(Node start, Node end, double maxLength) {
+    List<List<Edge>> allPaths = new ArrayList<>();
+    Queue<List<Edge>> queue = new LinkedList<>();
+
+    for (var edge : start.getEdges()) {
+      if (edge.length <= maxLength) {
+        queue.add(List.of(edge));
+      }
+    }
+
+    while (!queue.isEmpty()) {
+      List<Edge> currentPath = queue.poll();
+      Node currentPoint = currentPath.getLast().to;
+
+      if (currentPoint == end) {
+        allPaths.add(new ArrayList<>(currentPath));
+      } else {
+        for (var edge : currentPoint.getEdges()) {
+          if (!currentPath.contains(edge) && !currentPath.contains(edge.reversed)) {
+            List<Edge> newPath = new ArrayList<>(currentPath);
+            newPath.add(edge);
+            if (getLength(newPath) <= maxLength && !hasPointAppearingMoreThanTwice(newPath)) {
+              queue.add(newPath);
+            }
+          }
+        }
+      }
+    }
+
+    allPaths.sort(Comparator.comparingDouble(LoopLineMerger::getLength));
+
+    return allPaths;
+  }
+
+  private void removeShortStubEdges() {
+    for (var node : output) {
+      for (var edge : List.copyOf(node.getEdges())) {
+        if (edge.length < minLength && (edge.from.getEdges().size() == 1 || edge.to.getEdges().size() == 1)) {
+          edge.remove();
+        }
+      }
+    }
+  }
```

Comment:
With the old system where we broke input linestrings into straight segments in the split function the graph would always look like this:

![image](https://github.com/user-attachments/assets/e05e5d32-94f7-41e7-a0f5-e8bb914f7883)

Circle are edge endpoints. So stubs had always one end which had no other edge attached.

However, with the new way of only breaking up input linestrings at shared nodes, we can have graphs that look like this:

![image](https://github.com/user-attachments/assets/523a8df8-d34b-49ac-9c43-49fa50dbb0ff)


So the stubs can have start and end point at the same node. Let me add therefore an extra check here to take account of those looped-stubs.

Reply:
Would be good to have some tests for those.

Other strange things that I saw in the data look like this:

![image](https://github.com/user-attachments/assets/ed17caef-86bb-404d-935b-aa98217c2983)

Here loop merging when starting on edge 5 will remove the connection between C and D and result in this:

![image](https://github.com/user-attachments/assets/7e7a5981-c885-47bc-a80e-1c9b8dabf60f)

How this works:

- start with edge 5 which goes from C to C (looped stub)
- find all paths from C to C, which is 5, 1-3, 3-1
- remove first edge of all but the shortest path, i.e. remove edges 1 and 3
-

Reply:
I added a condition to removeLoops now which skips searching paths if the `edge.from == edge.to`...